docs/dev/core/new-metric-type.md

```diff
@@ -0,0 +1,212 @@
+                _ => Metric::Counter(amount),
+            })
+    }
+}
+```
+
+Use `glean.storage().record()` to record a fixed value or `glean.storage.record_with()` to construct a new value from the currently stored one.
+
+The storage operation makes use of the metric's variant of the `Metric` enumeration.
+
+## The `Metric` enumeration
+
+Persistence and in-memory serialization as well as ping payload serialization are handled through the `Metric` enumeration.
+This is defined in `glean-core/src/metrics/mod.rs`.
+Variants of this enumeration are used in the storage implementation of the metric type.
+
+To add a new metric type, add a new variant to the `Metric` enum:
+
+```rust,noplaypen
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Metric {
+    // ...
+    Counter(i32),
+}
+```
+
+Then modify the below implementation and define the right category name for the new type. This will be used in the ping payload:
+
+```rust,noplaypen
+impl Metric {
+    pub fn category(&self) -> &'static str {
+        match self {
+            // ...
+            Metric::Counter(_) => "counter",
+        }
+    }
+}
+```
+
+Finally, define the ping payload serialization (as JSON).
+In the simple cases where the in-memory representation maps to its JSON representation it is enough to call the `json!` macro.
+
+```rust,noplaypen
+impl Metric { // same block as above
+    pub fn as_json(&self) -> JsonValue {
+        match self {
+            // ...
+            Metric::Counter(c) => json!(c),
+        }
+    }
+}
+```
+
+For more complex serialization consider implementing serialization logic as a function returning a [`serde_json::Value`](https://docs.rs/serde_json/*/serde_json/enum.Value.html)
+or another object that can be serialized.
+
+## FFI layer
+
+In order to use a new metric type over the FFI layer, it needs implementations in the FFI component and the platform-part.
+
+### FFI component
+
+The FFI component implementation can be found in `glean-core/ffi/src/lib.rs`.
+
+Add a global map for your metric type:
+
+```rust,noplaypen
+lazy_static! {
+    static ref COUNTER_METRICS: ConcurrentHandleMap<CounterMetric> = ConcurrentHandleMap::new();
+}
+```
+
+Add a function to create new instances of this metric type:
+
+```rust,noplaypen
+#[no_mangle]
+pub extern "C" fn glean_new_counter_metric(
+    category: FfiStr,
+    name: FfiStr,
+    send_in_pings: RawStringArray,
+    send_in_pings_len: i32,
+    lifetime: i32,
+    disabled: u8,
+) -> u64 {
+    COUNTER_METRICS.insert_with_log(|| {
+        let send_in_pings = unsafe { from_raw_string_array(send_in_pings, send_in_pings_len) };
+        let lifetime = Lifetime::try_from(lifetime)
+            .map_err(|_| {
+                log::error!(
+                    "[Counter] Failed to convert from lifetime value {}",
+                    lifetime
+                );
+            })
+            .unwrap_or(Lifetime::Ping);
```

Comment:
An aside (not for this PR): Is there a way to encapsulate this to not duplicate it in every metric type constructor?

Reply:
Yes. Actually we can use `from` and default to `Lifetime::Ping`. I'll change it here and add another PR changing the code.